File: ptern-typescript/src/semantic/bounds.ts
import type { Atom, Definition, Exclusion, Expression, RangeItem, Repetition } from "../parser/ast";
import { decodedLength } from "./validator";

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export type Bounds = { min: number; max: number | null };

export function computePternBounds(parsed: {
  definitions: Definition[];
  body: Expression;
}): Bounds {
  const defBounds = computeDefBoundsAll(parsed.definitions);
  return computeExpressionBounds(parsed.body, defBounds);
}

// ---------------------------------------------------------------------------
// Definition bounds (memoised)
// ---------------------------------------------------------------------------

function computeDefBoundsAll(defs: Definition[]): Map<string, Bounds> {
  const defExprs = new Map<string, Expression>(defs.map(d => [d.name, d.body]));
  const acc = new Map<string, Bounds>();
  for (const def of defs) {
    computeDefBoundsMemo(def.name, defExprs, acc);
  }
  return acc;
}

function computeDefBoundsMemo(
  name: string,
  defExprs: Map<string, Expression>,
  acc: Map<string, Bounds>,
): void {
  if (acc.has(name)) return;
  const body = defExprs.get(name);
  if (body === undefined) return;
  const bounds = computeExpressionBounds(body, acc);
  acc.set(name, bounds);
}

// ---------------------------------------------------------------------------
// Expression bounds
// ---------------------------------------------------------------------------

function computeExpressionBounds(expr: Expression, defs: Map<string, Bounds>): Bounds {
  const seqs = expr.alternatives;
  if (seqs.length === 0) return { min: 0, max: 0 };
  const first = computeSequenceBounds(seqs[0]!, defs);
  return seqs.slice(1).reduce((acc, seq) => {
    const b = computeSequenceBounds(seq, defs);
    return { min: Math.min(acc.min, b.min), max: maxOpt(acc.max, b.max) };
  }, first);
}

function computeSequenceBounds(seq: { items: { inner: Repetition }[] }, defs: Map<string, Bounds>): Bounds {
  return seq.items.reduce<Bounds>(
    (acc, cap) => {
      const b = computeRepetitionBounds(cap.inner, defs);
      return { min: acc.min + b.min, max: addOpt(acc.max, b.max) };
    },
    { min: 0, max: 0 },
  );
}

function computeRepetitionBounds(rep: Repetition, defs: Map<string, Bounds>): Bounds {
  const inner = computeExclusionBounds(rep.inner, defs);
  if (rep.count === null) return inner;
  const { min, max: repMax } = rep.count;
  if (repMax.kind === "exact") return { min: inner.min * min, max: mulOpt(inner.max, repMax.value) };
  if (repMax.kind === "none") return { min: inner.min * min, max: mulOpt(inner.max, min) };
  // unbounded
  return { min: inner.min * min, max: null };
}

function computeExclusionBounds(excl: Exclusion, defs: Map<string, Bounds>): Bounds {
  return computeRangeItemBounds(excl.base, defs);
}

function computeRangeItemBounds(item: RangeItem, defs: Map<string, Bounds>): Bounds {
  if (item.kind === "charRange") return { min: 1, max: 1 };
  return computeAtomBounds(item.atom, defs);
}

function computeAtomBounds(atom: Atom, defs: Map<string, Bounds>): Bounds {
  switch (atom.kind) {
    case "literal": {
      const len = decodedLength(atom.content);
      return { min: len, max: len };
    }
    case "charClass":
      return { min: 1, max: 1 };
    case "interpolation":
      return defs.get(atom.name) ?? { min: 0, max: 0 };
    case "group":
      return computeExpressionBounds(atom.inner, defs);
    case "positionAssertion":
      return { min: 0, max: 0 };
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function addOpt(a: number | null, b: number | null): number | null {
  if (a === null || b === null) return null;
  return a + b;
}

function maxOpt(a: number | null, b: number | null): number | null {
  if (a === null || b === null) return null;
  return Math.max(a, b);
}

function mulOpt(a: number | null, n: number): number | null {
  if (a === null) return null;
  return a * n;
}

File: ptern-kotlin/src/main/kotlin/io/ptern/semantic/Bounds.kt
package io.ptern.semantic

import io.ptern.parser.ast.*

data class Bounds(val min: Int, val max: Int?)

internal fun computePternBounds(parsed: ParsedPtern): Bounds {
    val defBounds = computeDefBoundsAll(parsed.definitions)
    return computeExpressionBounds(parsed.body, defBounds)
}

// ---------------------------------------------------------------------------
// Definition bounds (memoised)
// ---------------------------------------------------------------------------

private fun computeDefBoundsAll(defs: List<Definition>): Map<String, Bounds> {
    val defExprs = defs.associate { it.name to it.body }
    val acc = mutableMapOf<String, Bounds>()
    for (def in defs) computeDefBoundsMemo(def.name, defExprs, acc)
    return acc
}

private fun computeDefBoundsMemo(
    name: String,
    defExprs: Map<String, Expression>,
    acc: MutableMap<String, Bounds>,
) {
    if (acc.containsKey(name)) return
    val body = defExprs[name] ?: return
    acc[name] = computeExpressionBounds(body, acc)
}

// ---------------------------------------------------------------------------
// Expression bounds
// ---------------------------------------------------------------------------

private fun computeExpressionBounds(expr: Expression, defs: Map<String, Bounds>): Bounds {
    val seqs = expr.alternatives
    if (seqs.isEmpty()) return Bounds(0, 0)
    val first = computeSequenceBounds(seqs[0], defs)
    return seqs.drop(1).fold(first) { acc, seq ->
        val b = computeSequenceBounds(seq, defs)
        Bounds(minOf(acc.min, b.min), maxOpt(acc.max, b.max))
    }
}

private fun computeSequenceBounds(seq: Sequence, defs: Map<String, Bounds>): Bounds =
    seq.items.fold(Bounds(0, 0)) { acc, cap ->
        val b = computeRepetitionBounds(cap.inner, defs)
        Bounds(acc.min + b.min, addOpt(acc.max, b.max))
    }

private fun computeRepetitionBounds(rep: Repetition, defs: Map<String, Bounds>): Bounds {
    val inner = computeExclusionBounds(rep.inner, defs)
    if (rep.count == null) return inner
    val min = rep.count.min
    val repMax = rep.count.max
    if (repMax is RepUpper.Exact) return Bounds(inner.min * min, mulOpt(inner.max, repMax.value))
    if (repMax is RepUpper.None) return Bounds(inner.min * min, mulOpt(inner.max, min))
    return Bounds(inner.min * min, null) // unbounded
}

private fun computeExclusionBounds(excl: Exclusion, defs: Map<String, Bounds>): Bounds =
    computeRangeItemBounds(excl.base, defs)

private fun computeRangeItemBounds(item: RangeItem, defs: Map<String, Bounds>): Bounds {
    if (item is RangeItem.CharRange) return Bounds(1, 1)
    if (item is RangeItem.SingleAtom) return computeAtomBounds(item.atom, defs)
    return Bounds(0, 0)
}

private fun computeAtomBounds(atom: Atom, defs: Map<String, Bounds>): Bounds = when (atom) {
    is Atom.Literal -> {
        val len = decodedLength(atom.content)
        Bounds(len, len)
    }
    is Atom.CharClass -> Bounds(1, 1)
    is Atom.Interpolation -> defs[atom.name] ?: Bounds(0, 0)
    is Atom.Group -> computeExpressionBounds(atom.inner, defs)
    is Atom.PositionAssertion -> Bounds(0, 0)
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

private fun addOpt(a: Int?, b: Int?): Int? = if (a == null || b == null) null else a + b
private fun maxOpt(a: Int?, b: Int?): Int? = if (a == null || b == null) null else maxOf(a, b)
private fun mulOpt(a: Int?, n: Int): Int? = if (a == null) null else a * n
